File: src/commands/backup.ts
import { hostname } from "node:os";
import { join, dirname } from "node:path";
import { resolveOutputDir } from "../utils/resolve-output";
import { getHome } from "../utils/home";
import { generateTimestamp } from "../utils/timestamp";
import { backupSources } from "../backup/sources";
import { scanContent, summarize, formatReport, applyRedactions } from "../scan";
import type { ScanResult } from "../scan";
import type { BackupEntry, BackupSource } from "../backup/types";

function parseArgs(args: string[]) {
  let redact = true;
  let archive = false;
  let outputDir: string | null = null;
  let only: string[] = [];
  let skip: string[] = [];

  for (let i = 0; i < args.length; i++) {
    if (args[i] === "--no-redact") redact = false;
    if (args[i] === "--archive") archive = true;
    if (args[i] === "-o" && args[i + 1]) outputDir = args[++i];
    if (args[i] === "--only" && args[i + 1]) only = args[++i].split(",");
    if (args[i] === "--skip" && args[i + 1]) skip = args[++i].split(",");
  }

  return { redact, archive, outputDir, only, skip };
}

function filterSources(sources: BackupSource[], only: string[], skip: string[]): BackupSource[] {
  let filtered = sources;
  if (only.length) filtered = filtered.filter((s) => only.includes(s.category));
  if (skip.length) filtered = filtered.filter((s) => !skip.includes(s.category));
  return filtered;
}

async function copyFile(
  entry: BackupEntry & { type: "file" },
  destRoot: string,
  redact: boolean,
  scanResults: ScanResult[],
): Promise<boolean> {
  const file = Bun.file(entry.src);
  if (!(await file.exists())) return false;

  let content = await file.text();
  const scanResult = scanContent(entry.dest, content);

  if (redact && scanResult.action === "skip") {
    scanResults.push(scanResult);
    return false;
  }

  if (redact && entry.redact) content = entry.redact(content);
  if (redact) content = applyRedactions(content, scanResult);

  scanResults.push(scanResult);

  const destPath = join(destRoot, entry.dest);
  await Bun.$`mkdir -p ${dirname(destPath)}`.quiet();
  await Bun.write(destPath, content);
  return true;
}

async function copyDir(entry: BackupEntry & { type: "dir" }, destRoot: string): Promise<number> {
  const glob = new Bun.Glob("**/*");
  let count = 0;

  try {
    for await (const relative of glob.scan({ cwd: entry.src, onlyFiles: true, dot: true })) {
      const srcPath = join(entry.src, relative);
      const destPath = join(destRoot, entry.dest, relative);
      await Bun.$`mkdir -p ${dirname(destPath)}`.quiet();
      const content = await Bun.file(srcPath).text();
      await Bun.write(destPath, content);
      count++;
    }
  } catch {
    // Directory doesn't exist or is unreadable — skip silently (tool may not be installed)
  }

  return count;
}

export async function backup(args: string[]) {
  const { redact, archive, outputDir, only, skip } = parseArgs(args);
  const resolvedOutput = await resolveOutputDir(outputDir);

  const ts = generateTimestamp();
  const backupDir = join(resolvedOutput, `backup-${hostname()}-${ts}`);

  const sources = filterSources(backupSources, only, skip);
  let totalFiles = 0;
  const backedUpCategories: string[] = [];
  const scanResults: ScanResult[] = [];

  for (const source of sources) {
    const entries = source.entries(getHome());
    let categoryCount = 0;

    for (const entry of entries) {
      if (entry.type === "file") {
        const copied = await copyFile(entry, backupDir, redact, scanResults);
        if (copied) categoryCount++;
      } else {
        categoryCount += await copyDir(entry, backupDir);
      }
    }

    if (categoryCount > 0) {
      backedUpCategories.push(`${source.category} (${categoryCount})`);
      totalFiles += categoryCount;
    }
  }

  if (totalFiles === 0) {
    console.log("No files found to backup.");
    return;
  }

  if (archive) {
    const archivePath = `${backupDir}.tar.gz`;
    const backupName = backupDir.split("/").pop()!;
    await Bun.$`tar czf ${archivePath} -C ${resolvedOutput} ${backupName}`.quiet();
    await Bun.$`rm -rf ${backupDir}`.quiet();
    console.log(`Archive saved to: ${archivePath}`);
    console.log(`  ${totalFiles} files across: ${backedUpCategories.join(", ")}`);
  } else {
    console.log(`Backup saved to: ${backupDir}`);
    console.log(`  ${totalFiles} files across: ${backedUpCategories.join(", ")}`);
  }

  if (redact) {
    const summary = summarize(scanResults);
    const report = formatReport(summary);
    if (report) console.log(report);
  }
}

File: internal/cli/backup.go
package cli

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"sort"
	"strings"
	"time"

	"github.com/doguyilmaz/dothaven/internal/backup"
	"github.com/doguyilmaz/dothaven/internal/registry"
	"github.com/doguyilmaz/dothaven/internal/scan"
	"github.com/doguyilmaz/dothaven/internal/sys"
	"github.com/doguyilmaz/dothaven/internal/tui"
	"github.com/spf13/cobra"
)

// backupGroups aggregates backup targets into selectable category groups
// (count per category, and whether any entry in it is high-sensitivity).
func backupGroups(targets []registry.BackupTarget) []tui.Group {
	type agg struct {
		count int
		enc   bool
	}
	m := map[string]*agg{}
	for _, t := range targets {
		a := m[t.Category]
		if a == nil {
			a = &agg{}
			m[t.Category] = a
		}
		a.count++
		if t.Sensitivity == registry.High {
			a.enc = true
		}
	}
	cats := make([]string, 0, len(m))
	for c := range m {
		cats = append(cats, c)
	}
	sort.Strings(cats)
	groups := make([]tui.Group, 0, len(cats))
	for _, c := range cats {
		groups = append(groups, tui.Group{Name: c, Count: m[c].count, Encrypted: m[c].enc})
	}
	return groups
}

// formatCategories renders a per-category count map as "shell (3), git (2)".
func formatCategories(perCat map[string]int) string {
	cats := make([]string, 0, len(perCat))
	for c := range perCat {
		cats = append(cats, c)
	}
	sort.Strings(cats)
	parts := make([]string, len(cats))
	for i, c := range cats {
		parts[i] = fmt.Sprintf("%s (%d)", c, perCat[c])
	}
	return strings.Join(parts, ", ")
}

func newBackupCmd(env *sys.OS) *cobra.Command {
	var noRedact, archive bool
	var output string
	var only, skip []string
	c := &cobra.Command{
		Use:   "backup",
		Short: "Copy tracked config files into a timestamped backup",
		Args:  cobra.NoArgs,
		RunE: func(cmd *cobra.Command, args []string) error {
			redact := !noRedact

			// A backup is the safety net before wiping a machine — remind the user
			// (on a terminal) that it captures saved config, not unsaved editor work.
			if stdoutIsTTY() {
				fmt.Println("Note: this captures saved config files — not unsaved editor buffers or")
				fmt.Println("in-memory state. Save and quit your editors before relying on it to wipe a machine.")
				fmt.Println()
			}

			// Interactive category picker: only on a terminal, and only when the
			// user hasn't already constrained the set with --only/--skip.
			if len(only) == 0 && len(skip) == 0 && tui.Interactive() {
				chosen, err := tui.SelectCategories("What to back up", backupGroups(registry.BackupTargets(env.Home(), registry.Entries)))
				if err != nil {
					return err
				}
				if len(chosen) == 0 {
					fmt.Println("Nothing selected.")
					return nil
				}
				only = chosen
			}

			dir := env.ResolveOutputDir(output)
			host, _ := os.Hostname()
			if host == "" {
				host = "machine"
			}
			backupDir := filepath.Join(dir, fmt.Sprintf("backup-%s-%s", host, sys.Timestamp(time.Now())))

			targets := registry.BackupTargets(env.Home(), registry.Entries)
			res, err := backup.Run(targets, backupDir, backup.Options{Redact: redact, Only: only, Skip: skip})
			if err != nil {
				return err
			}
			if res.TotalFiles == 0 {
				fmt.Println("No files found to backup.")
				return nil
			}

			// Write a self-describing MANIFEST into the tree (before archiving, so
			// it travels with the backup) — what's inside, what was excluded, and
			// how to restore it.
			manifest := backup.Manifest(backup.ManifestMeta{
				Host:     host,
				OS:       runtime.GOOS,
				Version:  cmd.Root().Version,
				Created:  time.Now().Format(time.RFC3339),
				Redacted: redact,
			}, res)
			if err := sys.WriteFileSecure(filepath.Join(backupDir, "MANIFEST.txt"), manifest); err != nil {
				return err
			}

			summary := formatCategories(res.PerCategory)
			if archive {
				archivePath := backupDir + ".tar.gz"
				tmp := archivePath + ".tmp"
				// runShell surfaces tar's exit code — env.Run tolerates non-zero
				// exit, which would let a partial archive pass and then delete the
				// only good copy. Write to a temp file, rename on success, and only
				// remove the source once a complete archive exists.
				if out, err := runShell(cmd.Context(), "tar", "czf", tmp, "-C", dir, filepath.Base(backupDir)); err != nil {
					_ = os.Remove(tmp)
					return fmt.Errorf("tar failed (backup kept at %s): %v: %s", backupDir, err, out)
				}
				if err := os.Rename(tmp, archivePath); err != nil {
					return err
				}
				_ = os.RemoveAll(backupDir)
				fmt.Printf("Archive saved to: %s\n  %d files across: %s\n", archivePath, res.TotalFiles, summary)
			} else {
				fmt.Printf("Backup saved to: %s\n  %d files across: %s\n", backupDir, res.TotalFiles, summary)
			}

			if len(res.SkippedSensitive) > 0 {
				fmt.Printf("\n⚠ %d high-sensitivity path(s) excluded from this plaintext backup:\n", len(res.SkippedSensitive))
				for _, d := range res.SkippedSensitive {
					fmt.Printf("    %s\n", d)
				}
				fmt.Println("  Carry these with: dothaven chezmoi-export --apply  (age-encrypted)")
			}

			if len(res.ReadErrors) > 0 {
				fmt.Fprintf(os.Stderr, "\n⚠ %d file(s) exist but could not be read — excluded from the backup:\n", len(res.ReadErrors))
				for _, d := range res.ReadErrors {
					fmt.Fprintf(os.Stderr, "    %s\n", d)
				}
			}

			if redact {
				if report := scan.FormatReport(scan.Summarize(res.ScanResults)); strings.TrimSpace(report) != "" {
					fmt.Println(report)
				}
			}
			return nil
		},
	}
	c.Flags().BoolVar(&noRedact, "no-redact", false, "keep raw values (skip secret redaction)")
	c.Flags().BoolVar(&archive, "archive", false, "create a .tar.gz instead of a directory")
	c.Flags().StringVarP(&output, "output", "o", "", "output directory (default: ./reports in a repo, else ~/Downloads)")
	c.Flags().StringSliceVar(&only, "only", nil, "only these categories (comma-separated)")
	c.Flags().StringSliceVar(&skip, "skip", nil, "skip these categories (comma-separated)")
	return c
}
